### ISSUE TITLE
tracking issue for default binding modes in match (RFC 2005, match_default_bindings)

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#42640]
This is a tracking issue for the "match ergonomics using default bindings mode" RFC (rust-lang/rfcs#2005).

**Status:** Awaiting stabilization PR and docs PR! [Mentoring instructions here](https://github.com/rust-lang/rust/issues/42640#issuecomment-374728162).

**Steps:**

- [x] Implement the RFC
    - [Initial mentoring instructions can be found here](https://github.com/rust-lang/rust/issues/42640#issuecomment-313535089)
    - [Help guide people to use it via a targeted error message.](https://github.com/rust-lang/rust/issues/45383)
- [x] [Decide to stabilize](https://github.com/rust-lang/rust/issues/42640#issuecomment-374727589)
- [x] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] How to handle the logic around explicit bindings and coercion? https://github.com/rust-lang/rust/issues/44848
- [ ] When precisely should constants "autoderef"? https://github.com/rust-lang/rust/issues/44849
- [x] Reset to by ref mode? https://github.com/rust-lang/rust/issues/46688

**Comment by nikomatsakis** [comment]
I'm actually not 100% sure the best way to implement this. It seems like we will need to add **adjustments** to patterns, and then integrate those into the various bits of the compiler. @eddyb, have any thoughts on that? You usually have some clever ideas in this respect. =)

**Comment by eddyb** [comment]
We should be able to treat adjustments more uniformly now, although it'd still be a lot of plumbing.

**Comment by crazymykl** [comment]
How does this interact with slice patterns (#23121)?

**Comment by nikomatsakis** [comment]
@crazymykl 

My expectation would be that, when we encounter a pattern like `[a, ..b]`, we check the default binding mode. If it is "ref", then `a` becomes `&slice[0]` and `b` becomes (effectively) `&slice[1..]`.

**Comment by nikomatsakis** [comment]
@eddyb do we want to use general purpose adjustments here, or something more limited (the current RFC, after all, only requires autoderef of `&` and `&mut` types). It seems like though if we *can* use adjustments, that'll lay us a nice foundation for the future potentially. But it may introduce a lot of questions (e.g., what does it mean to "unsize" a pattern) that are better left unasked.

**Comment by eddyb** [comment]
@nikomatsakis Autoderef or autoref? If it's one bit per pattern making it separate for now is fine.

**Comment by nikomatsakis** [comment]
@eddyb 

>  Autoderef or autoref? If it's one bit per pattern making it separate for now is fine.

autoderef -- that is, where you have a pattern like `Some`, the type you are matching might now be `&Some` or `&mut Some` (or `&&Some`, etc). I think I agree, I'm inclined to introduce a `PatternAdjustment` struct that just includes auto-deref and go from there.

**Comment by nikomatsakis** [comment]
Here is a rough and incomplete implementation plan for the match ergonomics RFC.
  
There are a few high-level things we have to do:

- figure out which patterns need to an "auto-deref" -- e.g., if you have `match &foo { Some(x) ... }`, then we want to record on the `Some(x)` pattern that there is an automatic "dereference", meaning that it is equivalent to `&Some(x)`
- figure out the "binding mode" for patterns, which will no longer always be explicit.

Right now, the code is setup to scrape this information directly from the "HIR" (the compiler's AST). The HIR node for a pattern is [`hir::Pat`](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L561-L566), of which the most interesting part is the `kind` field [of type `PatKind`](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L643-L679). If you have a pattern like `Some(x)`, then, that would be represented as a tree:

- [`TupleStruct`](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L659)
    - containing a vec of length 1, which is a [Binding](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L650)
    - the `Binding` would have a [BindingMode](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L631-L635) of `BindByValue`.
    
We want to make this equivalent to `&Some(ref x)`, which would be encoded as a:

- [Ref](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L671) pattern, encoding the `&`
  - TupleStruct, encoding the `Some`
    - Binding with `BindingMode` of `ref`, encoding the `ref x`

We don't however have enough information to do this when we construct the HIR, since we need type checking results to do it, so we can't change the HIR itself. The way we typically handle this sort of thing then is to have the typeck encode "side tables" with auxiliary information. These tables are stored in [`TypeckTables`](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/ty/context.rs#L207-L268) and they encode all kinds of information.

In this case, I think we want two tables:

- `pat_adjustments`, sort of roughly analogous to the existing [adjustments table](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/ty/context.rs#L224). It would have a type like `NodeMap<usize>`, I think.
    - The key of such a map is the "id" of the `Pat` -- stored in [the `id` field of the `Pat` struct](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/hir/mod.rs#L563).
    - The value would just be a number indicating how many implicit `&` patterns we insert before this pattern. So, in our example, the `Some` pattern would wind up with a value of 1. For the other patterns, we'd probably just have no entry, meaning "none". (You could also store 0, but why waste the memory.)
- `pat_binding_modes` would have type `NodeMap<hir::BindingMode>`. It would, for each binding pattern, indicating the actual binding mode -- this may vary from what is found in the HIR, since we may be encoding a `ref` and so forth. We may want to change the HIR then to have some different type, or perhaps an `Option<hir::BindingMode>` -- where `None` means that nothing was explicitly written, which is the normal case -- so as to make it more obvious that when the user writes `x` this does not mean a "by value" binding mode anymore.

**Probably a decent first PR is just to introduce the second table (`pat_binding_modes`) and rewrite all the existing code to use it.** Right now, code that wants to find the binding mode of a binding extracts the value straight from the HIR, as you can see in the following examples:

- the [check_match code](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc_const_eval/check_match.rs#L262-L285), which enforces various sanity checks.
- [`determine_pat_move_mode`](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc/middle/expr_use_visitor.rs#L792-L811), which scrapes info from the HIR and also checks the type of the values being matched by value to decide if this is a copy-or-move;
- [`local_binding_mode`, a helper in borrowck](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc_borrowck/borrowck/mod.rs#L874-L884) -- this uses the "HIR Map" to lookup, given the id, the node for a given binding, and extracts its binding mode; this would be rewritten to use the new table
- the [HAIR conversion code](https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc_const_eval/pattern.rs#L373-L382); this is a precursor to MIR construction, which constructs a lowered form of the patterns called the HAIR. The HAIR is intended to be a copy of the HIR taking into account all of the information encoded in various side-tables, so you don't want to change the HAIR itself, just this code which creates it by scraping the HIR.

This is not a comprehensive list, but it does have the major use-sites. You can get a more comprehensive list by doing `rg 'BindByValue|BindByRef'`, which is what I did.

OK, no time for more, but hopefully this helps somebody get started! Please leave a note if you are interested in taking this on, and feel free to ping me on IRC (`nmatsakis`) or gitter (`nikomatsakis`) with any questions (or ask in #rustc).

**Comment by tschottdorf** [comment]
I'll look into the first step:

> Probably a decent first PR is just to introduce the second table (pat_binding_modes) and rewrite all the existing code to use it.

**Comment by nikomatsakis** [comment]
@tschottdorf woohoo!

**Comment by tschottdorf** [comment]
Starting to look at this again. What do you think is a good next step, @nikomatsakis? I was considering another "plumbing" PR (introduce `pat_adjustments` which is always trivial), but that doesn't seem useful since it won't really be exercised. I'll try to get a handle on what the computations that populate the typeck tables look like and where they'll live. I assume you'll have more pointers at some point!

**Comment by tschottdorf** [comment]
Seems that the changes that populate both tables are going to happen in the general vicinity of `typeck`s [match handling][match], correct? Putting some of that logic in (and testing it) could be a good way to get started, even if it doesn't populate the tables yet.

Am I correctly assuming that the major (additional) place where these tables would be used is during [HIR->HAIR lowering][hair] so that the HAIR representation spells out the information in the tables explicitly (at which point we're "done" with the tables)?

[match]: https://github.com/rust-lang/rust/blob/7f397bdb062fe13a4707219a2f32486c5294f642/src/librustc_typeck/check/_match.rs#L119
[hair]: https://github.com/rust-lang/rust/blob/cd72f2e269357ea4da4d0cf5502af35f4f9dab9e/src/librustc_const_eval/pattern.rs#L373-L382

**Comment by tschottdorf** [comment]
Is the below correct (not sure if it's useful to have this method except for assertions, but it'll help me understand what's a non-referential type).

```rust
#[allow(dead_code)]
impl PatKind {
    /// Returns true if the pattern is a reference pattern. A reference pattern
    /// is any pattern which can match a reference without coercion. Reference
    /// patterns include bindings, wildcards (_), consts of reference types, and
    /// patterns beginning with & or &mut. All other patterns are non-reference
    /// patterns.
    ///
    /// See https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md#definitions
    /// for rationale.
    fn is_reference_pattern(&self) -> bool {
        // NB: intentionally don't use a catchall arm because it's good to be
        // forced to consider the below when adding/changing `PatKind`.
        //
        // FIXME: is the below correct? In particular, where do "consts of reference types"
        // end up?
        match *self {
            PatKind::Wild |
            PatKind::Binding(_, _, _, _) |
            PatKind::Ref(_, _) => true,
            PatKind::Struct(_, _, _) |
            PatKind::TupleStruct(_, _, _) |
            PatKind::Path(_) |
            PatKind::Tuple(_, _) |
            PatKind::Box(_) |
            PatKind::Lit(_) |
            PatKind::Range(_, _, _) |
            PatKind::Slice(_, _, _) => false,
        }
    }
}
```

**Comment by eddyb** [comment]
In rust, names are usually "paths", because if you can say `FOO` you can also say `std::i32::MIN`.
So constants would be `PatKind::Path`.

**Comment by tschottdorf** [comment]
@eddyb gotcha. So for `PatKind::Path(_)` we'd have to look into the `_` and (after some magic incantation) see if it's a const ref.

I still must be missing something. A "const of reference type" would be `const CONST_OF_REF_TYPE: &u8 = &5`? I don't even know how to use this in a pattern at all. Is there a simple example?

**Comment by eddyb** [comment]
You'd literally use that name in a pattern and it's equivalent to the pattern `&5`. What you want though is to look at the *type*, not the *shape* of the pattern, if you want to know whether it's a reference or not.

For coercions in expressions, we compute a type for an expression, then compare it with the "expected type" coming from the parent expression and if they don't match, we can perform some adjustments. I'd expect patterns to follow a similar route.

**Comment by tschottdorf** [comment]
I assume you mean the below, but I've tried and failed to get an example that compiles. What am I missing?

```rust
const CONST_REF: &i64 = &5;

fn main() {
    print!("{}", CONST_REF);
    let f = &5i64;
    match f {
        CONST_REF => (),
        _ => (),
    };
}

```

```
error[E0080]: constant evaluation error
 --> src/main.rs:1:25
  |
1 | const CONST_REF: &i64 = &5;
  |                         ^^ unimplemented constant expression: address operator
  |
note: for pattern here
 --> src/main.rs:7:9
  |
7 |         CONST_REF => (),
  |         ^^^^^^^^^
```

**Comment by eddyb** [comment]
Oh, odd. Well, a `&str` literal should work nevertheless.

**Comment by tschottdorf** [comment]
Ack, `&str` works, thanks. The fact that my example above doesn't compile is a bug, then? Should I file it?

**Comment by tschottdorf** [comment]
Wait a minute, it doesn't work (same error):

```rust
const CONST_REF: &str = &"foo";

fn main() {
    print!("{}", CONST_REF);
    let f = "foo";
    match f {
        CONST_REF => (),
        _ => (),
    };
}
```

```rust
const CONST_REF: &[u8; 2] = &[1u8, 2u8];

fn main() {
    let f: &[u8; 2] = &[2u8, 3u8];
    match f {
        CONST_REF => (),
        _ => (),
    };
}
```

**Comment by eddyb** [comment]
I meant `b"foo"` literals. Looks like `&expr` is not implemented in `rustc_const_eval`.

**Comment by tschottdorf** [comment]
Ah, gotcha. The below works.

```rust
const CONST_REF: &[u8; 3] = b"foo";

fn main() {
    let f = b"bar";
    match f {
        CONST_REF => (),
        _ => (),
    };
}
```

**Comment by tschottdorf** [comment]
I spelunked a little bit and made a few (completely nonfunctional, except possibly compiling) baby steps here: https://github.com/rust-lang/rust/compare/master...tschottdorf:pat_adjustments

They're probably a good indication of where I need help.

**Comment by tschottdorf** [comment]
Just to update the main thread, I'm looking at this again this week with the intent of having some of the examples in the RFC compile (and the rest of the compiler still working). Right now I'm blocked on (presumably) my unfamiliarity with the type system -- in the simplest example, due to [some code] in `check_match`, I see expected type `TyInfer(..)` in `check_pat_arg` where I would really hope to see `TyRef(..)` -- but @nikomatsakis will hopefully be able to get me unstuck.

[some code]: https://github.com/rust-lang/rust/blob/07d950f38f1719e2a54d9a515178b0e11dbaf108/src/librustc_typeck/check/_match.rs#L464-L473

Until I have anything sensible, I'll keep the discussion in https://github.com/tschottdorf/rust/pull/1.

**Comment by nikomatsakis** [comment]
Update: @tschottdorf has this almost working!

**Comment by nikomatsakis** [comment]
@tschottdorf As I was saying on Gitter, one thing missing from your branch is that you need to update the [mem-categorization code](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1097-L1098), which sadly has not yet been ported to operate on HAIR but rather works directly on HIR (that would actually be a nice refactoring, but anyway).

What that code tries to do is to generate a ["cmt"](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L179-L189), which is a representation of the path that is being referenced when something is borrowed or moved. We need to adjust the code to take the new inferred `&` patterns into account.

As [this comment](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1100-L1143) tries to explain, the relationship between a **pattern** and a **cmt** is somewhat inverted. That is, if you have a match like this:

```rust
let foo: &&(u32,);
match foo {
  &&(x,) => ...
}
```

The memory from which `x` is being extracted is reachable by the (fully explicit) path: `(**x).0`, which corresponds to a cmt like:

    field0 { deref { deref { foo } } }

though I think in the comment I use a distinct, somewhat weirder notation: `x->&->&.0`, where `->&` means "deref a `&` reference" (actually, in the comment I wrote `->@`, which dates back from the days when we had `Rc<T>` built-in and called `@T`, but anyway).

The key point here is that the **pattern nesting** looks the opposite, with the derefs from the outside:

    deref { deref { field0 { foo } } }

This is why the [code that "categorizes" ref patterns (and box patterns)](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1226-L1232) works the way it does:

- It starts with `cmt`, which is the categorization of the thing being matched (here, `foo`).
- It then [produces the cmt for a deref of that](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1230), yielding `deref { foo }`.
- It then passes *that* cmt in for the inner patterns (`deref { field0 { foo } }`), which will unspool into the final result of `field0 { deref { deref { foo } }` as desired.

So, to insert "false" derefs, I think we need to do something rather the opposite of the HAIR lowering -- that is, the HAIR lowering wrapped the pattern in extra derefs, but we need to go *before* the main code to insert *our* extra derefs. Roughly [here](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1148). Also, this code is a bit weird in that it doens't *produce* a result, it [invokes a callback](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/mem_categorization.rs#L1147) with the result as it goes. So it'll take a touch of tweaking. I have to run now but can try to sketch out the pseudo-code later on.

**Comment by Michael-F-Bryan** [comment]
Just out of curiosity, is it possible to generalise this behaviour to all patterns and not just `match` statements? 

I originally posted on the [internal forum] about a case where I encountered a similar problem the `default binding modes` is trying to solve, but on a `for` loop.

Excerpt from the forum post:

> My particular example is that I'm iterating over a `Vec<(usize, Bot)>` and want to update each `Bot`. Currently I've got to write a big mess of `&mut` and `ref` to make the compiler happy:
>
> ```rust
> for &mut (tok, ref mut bot) in &mut self.bots {
>   let actions = bot.tick(&self.game_state);
>    debug!("Bot {} executed {:?}", tok, actions);
> }
> ```
>
> Whereas it'd be much easier (and arguably just as understandable) if pattern matching could elide away the boilerplate so I can write something like this:
>
> ```rust
> for (tok, mut bot) in &mut self.bots {
>     let actions = bot.tick(&self.game_state);
>     debug!("Bot {} executed {:?}", tok, actions);
> }
> ```
> 
> In theory the compiler should be able to automatically dereference the tuple reference, then that because `mut bot` is trying to use a mutably borrowed thing it'll automatically insert `ref mut bot`.

[internal forum]: https://internals.rust-lang.org/t/making-patterns-more-ergonomic/5949

**Comment by tschottdorf** [comment]
This PR applies to all bindings, so I think [the impl PR](https://github.com/rust-lang/rust/pull/44614) should help you already!

See the below snipped, which [doesn't compile today](https://play.rust-lang.org/?gist=d344d9b93fe2ab257f8543e3d871e734&version=nightly) but does on #44614:

```rust
pub fn main() {
    let tuples = vec![(0u8, 1u8)];
    for (m, n) in &tuples { // (m, n) are bound like &(ref m, ref n)
        let _: &u8 = m;
        println!("{} {}", m, n);
    }
}
```

**Comment by Michael-F-Bryan** [comment]
That's awesome! I'm assuming it won't matter if the "inner" things (in this case `m` and `n`) are referenced mutably, or if one is borrowed mutably but the other is borrowed immutably, will it?

**Comment by tschottdorf** [comment]
The mutability translates as outlined in the RFC. For this example, `(m, mut n)` would bind as `&(ref m, ref mut n)`.

**Comment by phaylon** [comment]
@tschottdorf Wouldn't it be `(m, n)` with both being `&mut`? Or was there some change in stategy?

**Comment by tschottdorf** [comment]
@phaylon are you thinking about `(m, n) in &mut tuples`? I'm not aware that the RFC suggests the behavior your mention, but as always I could be wrong.

What's definitely correct is that my PR currently messes up the binding when I try it with `(m, mut n) = &mut tuples` (it binds `n` as a value). I'll figure out what's going on there (though not now, probably EOW).

**Comment by phaylon** [comment]
@tschottdorf For example:
```
let (l, r) = &mut (23, 42);
```
From my reading of the RFC I'd expect `l` and `r` to be `&mut i32`.

I might just be looking at the wrong place. Could you point me towards where `mut var` bindings implying `&mut` is specified?

**Comment by tschottdorf** [comment]
Oh, now I see where you're coming from. I didn't consider that you would have to have the `&mut` on the right hand side. Yes, you're correct that then the above example would desugar from

```rust
pub fn main() {
    let mut tuples = vec![(0u8, 1u8)];
    for (m, n) in &mut tuples {
        *m = 6;
        *n = 7;
        println!("{} {}", m, n);
    }
}
```

to

```rust
pub fn main() {
    let mut tuples = vec![(0u8, 1u8)];
    for &mut (ref mut m, ref mut n) in &mut tuples {
        *m = 6;
        *n = 7;
        println!("{} {}", m, n);
    }
}
```

and yes, that's the case. I confused myself above because the error message that example was giving me was unhelpful.

**Comment by arielb1** [comment]
## Coercion chicken

@nikomatsakis

Now that I'm reading this again:
In HAIR/MIR, the soundness of the situation is quite simple: coercions and subtyping occur only on vexprs. Subtyping can operate on *references within a vexpr*, but it's still at heart a vexpr operation. Match expressions do their subtyping when they bind the (potentially references) to the fields of the scrutinee into the bindings.

If we just wanted to be consistent with that, then the #23116 example:
``` rust
#![allow(dead_code)]
use std::fmt::Debug;
struct S(Box<Debug + 'static>);
impl S {
    fn bar<'a>(&'a mut self)->&'a mut Box<Debug + 'a> {
        match self.0 { ref mut x => x } // should not compile, but does
    }
}
fn main() {}
```

Would perform a lexpr->vexpr->lexpr conversion, and therefore a temporary, creating MIR as follows:
```
temp = self.0; // lexpr -> vexpr
scrutinee_temp = &mut temp; // vexpr -> lexpr + coercion
x = &mut (*scrutinee_temp); // match binding
ret_ptr = x; // return value assignment
return
```

Which would of course cause a "borrow does not live long enough" when it catches you trying to return a borrow of the local `temp`.

Obviously, creating such coercion temporaries by default will annoy anyone who tries to use `ref` patterns to actually match parts of the scrutinee value by reference, so we don't. If we see a `ref` pattern, we prevent coercions and the resulting round-trips.

However, if all the `ref` bindings occur behind a reference, creating the temporary can't actually be annoying in this way - you don't take references to the value, so the temporary is invisible and we don't need to avoid it. Therefore, avoiding coercions exactly when there are explicit `ref` patterns is a implementable and non-annoying strategy.

### An aside

The above description is not actually implementation-coherent with typeck - to satisfy closure inference, match bindings can't use subtyping, so rustc actually sometimes performs subtyping on immutable lexprs - see this comment:
https://github.com/rust-lang/rust/blob/dcb4378e18571fa01e20ef63820d960f1c2cc865/src/librustc_typeck/check/_match.rs#L331-L379

Note that typeck's strategy is imperfect and leads to spurious errors in some situations
```Rust
fn foo<'x>(mut x: (&'x isize, ())) {
    let a = 1;
    let (mut _z, ref _y) = x;
    _z = &a; //~ ERROR no subtyping for you!
}
      
fn main() {}
```

We'll also have to solve the closure inference wonkyness when we get to MIR-based regionck, but hopefully the MIR-based regionck won't have wonky anyway.

### Pure trouble

While this argument works today, it will put us in a somewhat sticky situation if we want to allow `DerefPure impls for newtypes:

```Rust
mod some_newtype {
    pub struct Newtype<T>(pub T);
    impl<T> Deref for Newtype<T> {
        type Target = T;
        fn deref(&self) -> &T { &self.0 }
    }
    impl<T> DerefMut for Newtype<T> {
        fn deref_mut(&mut self) -> &mut T { &mut self.0 }
    }
    unsafe impl<T> DerefPure for Newtype<T> { /* ... */ }
}

use some_newtype::Newtype;
fn main() {
    let n = Newtype(("hello",));
    {
        let s = "hi there".to_string();
        let (mut d,) = n;
        *d = &s; // this can't successfully modify `n`, so, create a
                 // temporary? fail? both options are unsatisfying
    }
}
```

**Comment by nikomatsakis** [comment]
I've opened https://github.com/rust-lang/rust/issues/44848 to track the so-called "coercion chicken" question.

**Comment by nikomatsakis** [comment]
I've also opened https://github.com/rust-lang/rust/issues/44849 to track another interesting question:

> Another interesting question that @tschottdorf encountered when implementing default binding modes: What do we do with constants? The RFC specifies that we ought to treat a `FOO` binding that resolves to a constant as something which can skip through `&T` types -- however, that runs into trouble if the type of the constant itself is `&str` or `&[T]`. The current logic at least skips through all `&T` or `&mut T` types when it skips through *any*, but handling `&str` correctly would require skipping through "only the right number". @tschottdorf implemented various rules but we should at minimum update the RFC to match.

**Comment by tschottdorf** [comment]
I can't be the only one who wonders: why 'coercion chicken'?

**Comment by nikomatsakis** [comment]
Another issue that arose from the implementation: how to handle constants of reference type? The RFC states that constants of reference type will not "autoderef", but that means that things like this don't work, and I tend to think that they should:

```rust
let s: &'static str = "abc";
match &s {
  "abc" => true,
  _ => panic!(),
};
```

**Comment by tschottdorf** [comment]
@nikomatsakis you can check the first box now.

**Comment by nikomatsakis** [comment]
Just filed https://github.com/rust-lang/rust/issues/46688, regarding a curious interaction that took me a bit to figure out. Probably a bug in the RFC. It has to do with what happens when you match a pattern like `(a, &b)` against a value of type `&(u32, &u32)`. When we skip the first `&`, we get into "ref by default" mode, but when we explicitly acknowledge the second one, we do *not* get back into "by value" mode. That's kind of annoying.

**Comment by sgrif** [comment]
This feature can cause some pretty nonsensical error messages right now. For example, if you accidentally shadow a unit struct, you get an error about this feature, even though this feature has nothing to do with it and the suggestion certainly wouldn't work:

```
	error[E0658]: non-reference pattern used to match a reference (see issue #42640)
  --> src/query_builder/delete_statement/mod.rs:107:13
   |
18 | if let Some(name) = params.get("name") {
   |             ^^^^ help: consider using a reference: `&name`
   |
   = help: add #![feature(match_default_bindings)] to the crate attributes to enable

error[E0308]: mismatched types
  --> src/query_builder/delete_statement/mod.rs:107:13
   |
18 | if let Some(name) = params.get("name") {
   |             ^^^^ expected str, found struct `schema::users::columns::name`
   |
   = note: expected type `str`
              found type `schema::users::columns::name`
```

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose that we **stabilize** this feature as currently implemented, though with one pending change. Here is a summary.

### Tests that document current semantics

The tests for this feature can be found in the following directories:

- [run-pass/rfc-2005-default-binding-mode](https://github.com/rust-lang/rust/tree/b1f8e6fb06d7362eeb2065347a7db94e76b1cb2f/src/test/run-pass/rfc-2005-default-binding-mode)
- [ui/rfc-2005-default-binding-mode](https://github.com/rust-lang/rust/tree/b1f8e6fb06d7362eeb2065347a7db94e76b1cb2f/src/test/ui/rfc-2005-default-binding-mode)

### Changed from the RFC

The one change from the RFC was to "reset" the binding mode to "by value" whenever you encounter an explicit `&` or `&mut` pattern. This is to resolve the confusion described in https://github.com/rust-lang/rust/issues/46688 (Note: not yet implemented, but I will have a pending PR opened soon).

### Questions encountered

We encountered two curious cases:

**Constants.** The first was how to treat constants (https://github.com/rust-lang/rust/issues/44849): what should the semantics be when matching a constant of reference type? The current solution (also what is described in the RFC, I believe) is to be conservative: such a pattern does not attempt to insert "auto-derefs" (or "auto-ref-patterns", as you prefer). This does imply that [the following does not compile](https://play.rust-lang.org/?gist=debdeb8b942d1b1f960588abf8885340&version=nightly):

```
#![feature(match_default_bindings)]

fn main() {
    let x = "123";
    match &x {
        "123" => println!("Yes!"),
        _ => println!("No!"),
    }
}
```

Note that `&"123"` will work. This doesn't seem that important, though, and **moreover it would be backwards compatible to fix it in some way (since it is now an error)**.

**Coercions.** The other problem was the interaction with coercions (#44848). Re-reading [this comment I wrote](https://github.com/rust-lang/rust/pull/44614#issuecomment-331542696), and @arielb1's replies, it seems like the current rule -- using the presence of a syntactic `ref` as a marker for when to disable coercion -- suffices. And anyway the worst that could happen is that we get less than ideal errors, basically.

**Comment by withoutboats** [comment]
I turned this feature on recently when I was destructing a syn AST by reference, and it was an absolute joy to use. :-)

**Comment by nikomatsakis** [comment]
@withoutboats 

> I turned this feature on recently when I was destructing a syn AST by reference, and it was an absolute joy to use. :-)

Heck yes. I can't live w/o it anymore.

**Comment by nox** [comment]
It seems that the RFC says that rustc may end up using a `ref mut` binding mode for values on which `&mut` methods are called, am I reading this wrong? Suggestions about making `let foo = …` be the same as `let mut foo = …` were met with much criticism, so I'm really confused now if I'm correct about my understanding of this RFC.

**Comment by petrochenkov** [comment]
>I turned this feature on recently when I was destructing a syn AST by reference, and it was an absolute joy to use. :-)

I used them in rustc where possible since they were implemented and found the result slightly harder to read and to figure out what happens with ownership.

**Comment by petrochenkov** [comment]
At least it certainly requires some time and "rewiring" in the brain to happen to read the code written in the new style.

**Comment by petrochenkov** [comment]
Some minor annoying detail: if the binding mode is inferred to be by-reference, then "small" values like integers are bound by reference as well and you have to add dereferences.
Example:
```rust
// Previously
Variant { ref my_vector_of_things, my_int } =>
    MyStruct { len: my_vector_of_things.len(), my_int }

// With default binding modes
// Ugly `*` and can't use field shortcut anymore
Variant { my_vector_of_things, my_int } =>
    MyStruct { len: my_vector_of_things.len(), my_int: *my_int }
```
https://internals.rust-lang.org/t/suggestion-references-in-struct-literal/6789/10 may help here though.

**Comment by nox** [comment]
I hope this RFC isn’t going to be stabilised any time soon, given no large project used it yet in anger, and given @petrochenkov mentioned earlier that he felt it was making ownership of values way less obvious.

Could someone write a tool to patch existing code to see what impact it would have on Servo?

I don’t think any of these changes should land without being tried out in Servo, after all this is why Servo doesn’t use stable.

**Comment by pcwalton** [comment]
This is similar to how `&`/`&mut` is inferred in closures. We used to have capture clauses and it turned out to not matter at all when they were eliminated. I don't really share the concerns.

**Comment by nox** [comment]
I do not see how this is similar, a closure will capture as `&mut` only if the value was already mutable or a `&mut`, it doesn't make bindings that are not mutable suddenly mutable.

**Comment by pcwalton** [comment]
Though I do agree that we shouldn't stabilize until big projects have used this.

**Comment by nikomatsakis** [comment]
> Some minor annoying detail: if the binding mode is inferred to be by-reference, then "small" values like integers are bound by reference as well and you have to add dereferences.

Yes, this was an anticipated outcome. Our planned way to address this was through https://github.com/rust-lang/rust/issues/44619 -- basically, by making it so that it matters less if you have a `&u32` vs a `u32` (that distinction, in general, is kind of annoying, and arises in a number of other contexts -- e.g., `for i in &slice { .. }`, or `slice.iter().filter(|i| ...)`).

**Comment by nikomatsakis** [comment]
@nox 

> It seems that the RFC says that rustc may end up using a `ref mut` binding mode for values on which `&mut` methods are called, am I reading this wrong? 

I believe you are. The binding mode does not depend on how the value is used: only on the type of the value being matched, and the pattern used to match it.

It is certainly possible though to get a `ref mut` binding mode:

```rust
fn foo(x: &mut Option<Vec<u32>>) {
    if let Some(v) = x {
        v.push(22); // modifies caller's vec
    }
}
```

Similar things are possible today. For example, if I swap the type of the input to `Option<&mut Vec<u32>>`:

```rust
fn foo(x: Option<&mut Vec<u32>>) {
    if let Some(v) = x {
        v.push(22); // modifies caller's vec
    }
}
```

In general, the operation of `match` is meant to be analogous to how `into_iter` works today. So indeed I can write the following function today, which uses the original argument type `&mut Option<Vec<u32>>`) but paired with a `for` loop, and also mutate the caller's vector in an analogous way:

```rust
fn foo(x: &mut Option<Vec<u32>>) {
    for v in x {
        v.push(22); // modifies caller's vec
    }
}
```

> given @petrochenkov mentioned earlier that he felt it was making ownership of values way less obvious

I believe that @petrochenkov's exact words were that they "found the result **slightly harder** to read and to figure out what happens with ownership" (emphasis mine). 

In general, it is true that you get somewhat less information from the patterns in terms of ownership -- but already that information could be misleading. For example, we saw above that a `Some(v)` pattern might mean that you own `v`, but it might also be that `v` is itself a borrowed reference (in which case, yes, you own the reference, but not the *data*).

I've found that this feature actually generally makes it *easier* for me to follow what's going on, because I can think more about "borrowed vs owned" and worry less about the particulars of the type. I quite like that `Option<&mut Vec<u32>>` and `&mut Option<Vec<u32>>` interact with matches in the same, analogous way, for example -- and I like having `match` work more like `for`. 

In terms of being used with large projects, usage in the wild is always good! The feature is certainly being used in rustc, I'd like to hear about other people's experiences with it of course. (Often though waiting for the biggest projects to adopt can be tricky, since people wait to use nightly features until they are on stable.)

**Comment by sgrif** [comment]
> Though I do agree that we shouldn't stabilize until big projects have used this.

Given that 100% of code written using this feature can be written without using this feature, I think it's a stretch to assume many large projects are going to use this.

I tried enabling this on Diesel's codebase, and replaced as much code as I could find (`rg "match \*"` is what I looked for) to take advantage of this feature. You can find the results here: https://github.com/diesel-rs/diesel/pull/1578

I encountered the issue that @petrochenkov described in https://github.com/rust-lang/rust/issues/42640#issuecomment-368163763 3 times. In all 3 cases, I opted out of this feature by continuing to `match *expr`. I did find it surprising that this occurred.

I would have thought that this feature would never infer `ref` for types which are `Copy`. While it's stumbling block that is smaller than our current situation, I do still think it's a stumbling block. While it's less common and less intrusive, I also think it's less clear why it occurred than the existing rules. Perhaps stabilizing this should at least wait until #44619 has been explored, and we're confident that we will be able to remove this gotcha backwards compatibly in the future.

That said, overall this enabled me to remove meaningless punctuation and `ref`s from the codebase. Other than the cases where I wanted a copy, this feature let me say what I mean. I disagree that this somehow obscures ownership semantics. `match *foo { Bar { ref baz } => qux(&**baz) }` is not exactly clear to begin with.

**Comment by scottmcm** [comment]
It seems like we'd always have other explicit options (brainstorming: `match foo { Bar { move baz } => ` or `match foo { Bar { baz: i32 } =>`), so I don't think we need to wait on #44619 here.

**Comment by nikomatsakis** [comment]
@sgrif  Thanks for that exploration! Very helpful.

> Perhaps stabilizing this should at least wait until #44619 has been explored, and we're confident that we will be able to remove this gotcha backwards compatibly in the future.

I was wondering about this. It might make sense -- that said, I don't particularly like the idea of having special rules around `Copy` types, particularly since just because something *is* `Copy` doesn't mean that you *want* to copy it (it could be large, or be something like a closure or copy type that is `Copy`). If we went that route, I think we would really *need* to have a distinct between "pod" (can be memcopied, but has its own identity and shouldn't be) and "copy" (no identity at all). We decided some time ago not to do that, for better or worse, and I think it'd be hard to back out from it now... though I can't precisely remember why we felt like this was a 1.0 issue that we couldn't change later... I guess because it would alter the trait hierarchy in ways that are not trivially backwards compat.

**Comment by leodasvacas** [comment]
Also afaik implementing `Copy` (and traits in general) is backwards compatible, we probably need autoref to be able to do stop inferring `ref` without breaking that property.

**Comment by nikomatsakis** [comment]
OK, there were a few concerns raised in the FCP. We discussed in the lang-team meeting and decided to go forward with the stabilization, but I wanted to record the concerns and the responses to them for posterity:

---

@nox [pointed out](https://github.com/rust-lang/rust/issues/42640#issuecomment-368159723) that this feature allows `ref mut` bindings when matching against `&mut` values (by design). This means that -- compared to today's matches -- there is "one less use of `mut`" to create a mutable reference that affects other variables. However, as I [described here](https://github.com/rust-lang/rust/issues/42640#issuecomment-369023962), there is precedent for this in `for` loops and `match` statements in other, similar scenarios.

Similarly, @petrochenkov noted that [using this feature sometimes makes it less clear where borrowing is happening](https://github.com/rust-lang/rust/issues/42640#issuecomment-368161767), but [others](https://github.com/rust-lang/rust/issues/42640#issuecomment-369026752) [reported](https://github.com/rust-lang/rust/issues/42640#issuecomment-369023962) opposite experiences.

---

@nox raised the concern that the feature has [not been used "in the wild" enough](https://github.com/rust-lang/rust/issues/42640#issuecomment-368217592), but there are a number of experience reports, [including an "at-scale" deployment on Diesel](https://github.com/rust-lang/rust/issues/42640#issuecomment-369026752), so we feel comfortable going forward.

--- 

@petrochenkov [pointed out](https://github.com/rust-lang/rust/issues/42640#issuecomment-368163763) that the implied ref bindings are annoying around copy types. This is certainly true, and was known in advance as a shortcoming of the RFC. However, that problem too arises in other contexts (e.g., `foo.iter()`), and the consensus from the RFC discussion was that the [right way to handle this](https://github.com/rust-lang/rust/issues/42640#issuecomment-369305245) is by tackling the `&i32 -> i32` problem head on. Moreover, [as @leodasvacas pointed out here](https://github.com/rust-lang/rust/issues/42640#issuecomment-372502464), having rules that depend on whether the matched value is `Copy` would interact poorly with semver.

---

Thanks all!

**Comment by nikomatsakis** [comment]
I'm adding the E-mentor tag... we are in need of a stabilization PR here! Instructions can be found [on forge](https://forge.rust-lang.org/stabilization-guide.html), and I'd be happy to help with any other concerns.

In addition, we will need to update documentation: cc @rust-lang/docs, any recommendations on what to do there?

**Comment by steveklabnik** [comment]
> In addition, we will need to update documentation: cc @rust-lang/docs, any recommendations on what to do there?

I'll be opening a pre-rfc in the next few hours; tl;dr don't worry about it. I'll make sure it won't get lost.

**Comment by Dylan-DPC** [comment]
@nikomatsakis HI. Can work on this :+1:

**Comment by nikomatsakis** [comment]
@Dylan-DPC great! You basically want to search through for the feature flag `match_default_bindings`, as [described here](https://forge.rust-lang.org/stabilization-guide.html).

**Comment by kennytm** [comment]
Note that there is still an ICE involving `match_default_bindings`: #46197.

**Comment by nikomatsakis** [comment]
Oh yeah, forgot about that. We should fix it.

**Comment by cramertj** [comment]
@Dylan-DPC Hey! Have you had a chance to look at this? 1.26 is supposed to branch soon, so if we want t get `match_default_bindings` as part of this release we need to have a PR open very soon. No pressure on you if you're too busy-- I just wanted to check in.

**Comment by ErichDonGubler** [comment]
@steveklabnik: Did you still want help with documenting this? I might be interested in contributing. :)

**Comment by Dylan-DPC** [comment]
@cramertj haven't started it yet. will do this week

**Comment by leodasvacas** [comment]
@cramertj Isn't it better to stabilize language features right after a new beta rather than right before? This way there is more time to fix potentially unnoticed flaws by testing it in the wider nightly ecosystem.

**Comment by cramertj** [comment]
@leodasvacas It is the case that the more time we have to spend with an unstable feature, the more time we have to find bugs in it. However, this approach is in opposition to our other goal of shipping features. This is maybe a bit of a blunt way to put it, but there's long been consensus that this feature is ready for stabilization, and I'd like to move forward on that basis.

**Comment by steveklabnik** [comment]
Did https://github.com/rust-lang/rust/pull/49394 close this issue?

**Comment by ErichDonGubler** [comment]
@steveklabnik: Looks like this was stabilized, but the as-yet-unchecked items in "Unresolved questions" were punted to be fixed after shipping binding modes first?

Source: https://github.com/rust-lang/rust/pull/49394#issuecomment-376326024 and the response.

**Comment by Boscop** [comment]
Before we stabilize `match_default_bindings`, can we talk about this? https://github.com/rust-lang/rust/issues/50008

**Comment by nikomatsakis** [comment]
@Boscop 

> Before we stabilize `match_default_bindings`, can we talk about this? #50008

This is the case of copy data -- and yes, the current plan, as discussed [in this summary](https://github.com/rust-lang/rust/issues/42640#issuecomment-374727589) -- is to improve this case via a coercion from `&usize` to `usize`. 

Your suggestion, I believe, is to make `&P` patterns, when machting a non-reference type *and* in a "default binding mode" of `ref`, go back to a default binding mode of  `move`. This way `for (&a, b) in &vec` would "copy out" from `a` but take `b` by ref; the fully explicit pattern would be `for &(a, ref b) in &vec`. I confess I have often tried to edit patterns in just the way you describe, due I think to habit in writing things like `for &i in &vec`. Still, it seems confusing (to me) to have `&` "reset' the default binding mode, rather than pairing off with a `&T` type -- and there is also the common belief among newcomers that `&` should mean what "ref" means to contend with (that is, when first reading your example, I thought you were attempting to do the opposite of what in fact you are attempting). It was also considered to add a `move` binding mode, which would allow one to write `for (move a, b) in &vec`, but that was confusing too.

So in short I still think that the coercion is the best approach.

**Comment by rpjohnst** [comment]
I am absolutely floored by the fact that default binding modes don't *already* work like @Boscop expects. That (apparently incorrect) understanding was part of why I came around to liking them.

Further, the `&usize` -> `usize` coercion was fairly controversial. It was kind of swept up in a larger umbrella issue https://github.com/rust-lang/rust/issues/44619, which notably does not include any work on implementing or experimenting with such a coercion. (It does include the less-controversial auto(de)ref on operators only, which already applies to methods, and the opposite coersion of autoref.)

So given that several of us keep running into the desire for `&` to "reset" the binding mode, I don't think the difference between that and "pairing with a `&T`" is all that confusing. You can easily think of default binding modes as transforming the types you're matching on, such that you *are* "paring with a `&T`." After all, a bare name *does* wind up having `&T` as its type.

Both "`&` resets the binding mode" and "auto(de)ref for operators" feel very economical- like they're filling in a hole and fulfilling a promise made by the language. Adding a more general `&usize` -> `usize` coercion doesn't. I would really like to see *at least* some experimentation with "`&` resets the binding mode."

**Comment by Boscop** [comment]
> Both "& resets the binding mode" and "auto(de)ref for operators" feel very economical- like they're filling in a hole and fulfilling a promise made by the language. Adding a more general &usize -> usize coercion doesn't. I would really like to see at least some experimentation with "& resets the binding mode."

100% agree with this. Also, if `&T` constituents are basically coerced into `T` if `T: Copy`, that would lead to inconsistencies when it's a `&mut T`. That one should not be auto coerced to `T` (because one might want to mutate it) but (when not mutating) it should still be possible to move it out explicitly by writing `&x` (instead of the more verbose `&mut x`. That would be the most economic and consistent / intuitive solution imo. (Using the move keyword here would feel very inconsistent and counter intuitive.)

**Comment by Boscop** [comment]
default match bindings are already magic and auto coercion would make it even more confusing for newbies than explicitly moving out.

> I confess I have often tried to edit patterns in just the way you describe, due I think to habit in writing things like for &i in &vec. Still, it seems confusing (to me) to have & "reset' the default binding mode, rather than pairing off with a &T type

It feels natural to try `&x` to move it out for experienced Rust users. 

Btw, in my experience, any kind of magic that makes code more concise for experienced users obfuscates the code more for newbies, because magic is by its nature a special-cased introduced inconsistency that has to be learned in addition to the already complex normal set of rules. This is very apparent when math professors use a math notation that uses a lot of syntactic sugar to be more concise, rather than explicit. 
When learning, it's better to see all the details spelled out to internalize them, before moving on to hide them under syntactic sugar. Otherwise the understanding will only be cloudy.

As someone who knows a complex subject (like Rust) it's very difficult to imagine how someone who doesn't yet know it processes the information because it's become so second nature..

But I remember from math classes in uni that people seemed to understand the concepts better without syntactic sugar.

So while I think there is some merit to magic like default match bindings, I would not recommend anyone to use this feature who hasn't yet 100% understood the whole borrow checker and the difference between `ref` and `&`. Imo it's mostly useful for experienced users who want to write more concise code, but it doesn't make (this part of) Rust easier to learn for newbies.

And I wish there was more A/B testing and experiments by supervising people who are learning Rust before making decisions to change the language with the goal of making it easier to learn to confirm that those changes actually make it easier to learn... 

So I think we should 1. recommend newbies not to use default match bindings but being explicit, 2. when using default match bindings, we should use `&x` to move out constituents because for experienced users it makes the most sense.

**Comment by eddyb** [comment]
I *really* like `&Ctor(...field_pattern)` and `Ctor(...&field_pattern)` being equivalent, which makes `&Ctor(...ref field_var)` and `Ctor(...&ref field_var)` (i.e. `Ctor(...field_var)`) equivalent.
Mixing subpatterns with and without `&`  chooses between "by-value" and "by-ref".
cc @nikomatsakis

**Comment by cramertj** [comment]
Yeah, in fact this is the behavior I initially thought @nikomatsakis had implemented [here](https://github.com/rust-lang/rust/pull/48448#issuecomment-367877306) but then realized that the PR had taken a more conservative approach. I'm totally in favor of this change.

**Comment by Boscop** [comment]
@eddyb Yes, that consistency is why `&x` would make a lot more sense than the `move` keyword, or not even allowing this.

@cramertj It seems in that thread you meant the same thing I was expressing above, which is another indication that this would be more intuitive / expected behavior for experienced Rust users.

**Comment by joshtriplett** [comment]
My apologies if this message is redundant; I'm trying to catch up on the current state of things and I think I've missed some context across a few different threads, so I'm seeking some clarification.

If I understand the current state correctly, matching on an expression that has a reference type will allow you to write a pattern that matches out what *looks* like a non-reference but automatically becomes a reference rather than giving a type error? For instance, you can match on `&Foo(expr)` with a pattern like `Foo(name)` and get out a `name` that's a reference rather than a value?

But matching an expression that has a reference type will never let you extract a non-reference from it, and vice versa, matching an expression that has a non-reference type will never let you extract a reference from it (without writing `ref`)?

Is that accurate? And is there an intention to change that, or are we currently committing to *not* changing that? In particular, I think I'm confused by the same comment from @nikomatsakis that @cramertj referenced.

If that's accurate, then while I still wish we didn't have this at *all*, at least that's only breaking my internal typechecker but not actually making it easier to write code that has unexpected costs or correctness concerns.

**Comment by petrochenkov** [comment]
Looks like `match_default_bindings` was released as is on stable channel just 3 hours ago (https://github.com/rust-lang/rust/pull/50510).

**Comment by withoutboats** [comment]
> But matching an expression that has a reference type will never let you extract a non-reference from it

This is not correct. It works as it always has - you destructure the reference with an `&Foo` pattern. Other people are talking about inserting a reference character where there isn't actually a reference type.

**Comment by joshtriplett** [comment]
@withoutboats Sorry, I should have phrased that more carefully. I meant "without specifying the `&`. So, for instance, you could match `&x` with a pattern of `&y` and end up with a non-reference `y`, but you couldn't match `&x` with a pattern of `y` and end up with a non-reference `y`. Right?

**Comment by bstrie** [comment]
Does the stabilization in the recent 1.26 release represent a complete implementation of RFC 2005? If not, what remains? If so, then shall we move discussion of further extensions into a new RFC?

**Comment by nikomatsakis** [comment]
@bstrie I believe this is all "new RFC" material, yes.

**Comment by nikomatsakis** [comment]
(And I think it is a backwards compatible extension)

**Comment by eddyb** [comment]
I'm not entirely sure this is backwards-compatible, because you *can* use `Some(&x)` currently to match `&Option<&T>` and it results in `x: T`, whereas the proposal would make `x: &T`.

**Comment by rpjohnst** [comment]
That's deeply unfortunate, how did we wind up with that? 😱

**Comment by cramertj** [comment]
I think we can avoid that by only performing the binding mode switch back to by-move when matching a non-reference with a reference pattern.

**Comment by cramertj** [comment]
Basically, just like what we did with the first iteration of default-match-binding-modes, we only change the case that doesn't compile at all today.

**Comment by joshtriplett** [comment]
@cramertj I'd like to suggest that "changing a case that doesn't compile to now compile" is not a heuristic we can or should apply universally. Sometimes it's the compiler's *job* to refuse to compile something and complain, so that code gets fixed.

In this particular case, if it works in some cases but not in others, and the demarcation between the two isn't extremely clear, or the behavior isn't consistent (which the case @eddyb mentioned suggests it wouldn't be), then that seems worse than not supporting it at all.

**Comment by leodasvacas** [comment]
@rpjohnst It's the exact behaviour [specified on the RFC](https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md#definitions).

**Comment by sgrif** [comment]
Is there any way to turn this off within a codebase? For libraries that support <1.26, this is going to be a pain point. CI ensures that we don't accidentally rely on it, but it'd be great if there were a way to make sure that builds fail locally for new contributors regardless of which Rust version they're using

**Comment by leodasvacas** [comment]
@rpjohnst You've convinced me, the implementation does not match the RFC specification. So either we specify the implementation and amend the RFC or we fix the implementation, which should be done sooner rather than later, perhaps even taking advantage of the possibility of a point release #50756.

**Comment by nikomatsakis** [comment]
@rpjohnst I believe that was the behavior we opted for in https://github.com/rust-lang/rust/issues/46688

**Comment by nikomatsakis** [comment]
@sgrif there is not presently a way to turn this (or other) features off that I know of; interesting idea though.

**Comment by rpjohnst** [comment]
Ah, so the extra rule is that `&` patterns no longer preserve the binding mode, but unconditionally reset it. I'd still rather revert that change- it's an extra rule to remember, and it was never in the RFC so it looks like it didn't get discussed much. :(

Though on the other hand, I'm not actually sure whether it makes @Boscop's proposal backwards-incompatible. Could you expand on that @eddyb? Matching a `&` against an actual `&` would have the existing meaning (switch to `move` *and* deref) while matching a `&` against a non-`&` while in `ref` mode would *just* switch to `move`.

On the gripping hand, I'm also strongly in agreement with @joshtriplett in this case. Giving `&` *both* those meanings seems very confusing. So I kind of feel like it should be either-or- do we want the existing behavior or do we want the new behavior? I really like the consistency of the new behavior (e.g. https://github.com/rust-lang/rust/issues/42640#issuecomment-386567098), and as I've said it's how I always expected it to work.

**Comment by eddyb** [comment]
@rpjohnst Maybe I inaccurately represented @Boscop's proposal - my version doesn't even have a "binding mode" - just that matching `&T` by a `T` pattern results in *extra* `&` pushed to sub-patterns.

**Comment by nikomatsakis** [comment]
I'm a bit torn on all this. On the one hand, I find @eddyb's idea interesting — though I think there are many variants and I'm not 100% sure what is being proposed. My understanding was roughly "if you have a value of type `&T` and a ctor for `T`, you can "propogate" the `&` inside". I like that idea but it has a lot of ramifications that are quite different from the existing proposal — and they go beyond #46688.

For example, if I am matching a value of type `&Option<&Option<T>>` with the pattern `Some(Some(x))`, the current master gives `x` the type `&T`. If I understood @eddyb's proposal, it would give `&&T` — or perhaps error, it's not entirely clear to me. In particular, we would begin by propagating the `&` through the first `Some(..)` pattern, giving a value of type `&&Option<T>`. Now we match that with `Some(..)`, which either errors or gives a value of type `&&T`.

So basically it seems to me that we are *really* talking about rolling back the whole match default bindings change. This is a much bigger thing than #46688.

(When I thought it was just #46688 that was under discussion, I was considering this to be rather similar to the ["course correction"](https://internals.rust-lang.org/t/pre-rfc-adjust-default-object-bounds/2199/31) that we took for Object Lifetime Default rules. The reasoning here is slightly difference, but ultimately I think that was the right call. But this seems to be a much larger proposal.)

Now, maybe there are other proposals where reverting `#46688` would be enough. Perhaps @Boscop's proposal, for example?

**Comment by eddyb** [comment]
@nikomatsakis I would expect an error and require `Some(&Some(x))`, which has always been a valid pattern *for the type behind the reference*, but the possibility of a valid `Some(&&Some(x))` pattern makes me uneasy and I begin to agree with the motivation for a more "deeper flattening" of references.

**Comment by nikomatsakis** [comment]
@eddyb the need to sometimes introduce dummy temps just to make "doubly indirect" references feels bad to me, if that's what you mean (i.e., if we were going to produce a `&&T` value for `x`)

**Comment by eddyb** [comment]
@nikomatsakis No, in my view, you would never add more than one reference to bindings, because you wouldn't add more than one reference to sub-patterns on a constructor matching a single level of reference to the type of the constructor.

**Comment by leodasvacas** [comment]
I see nothing wrong with the current rules, and certainly see no motive to rollback of stabilized functionality.

In fact I like the extension that makes strictly more things compile, it's a DWIM approach. The interpretation is that a `&` in a pattern always means by-move. If it's matching a `&T`, then move `T`. If it's matching `T` in ref binding mode, then move `T`.

**Comment by eddyb** [comment]
@leodasvacas I agree with that sentiment, now that I've seen examples with nested references.
My idea was "interesting" in the very strict sense that it had a "bijective" sugar, but nothing more.

**Comment by Boscop** [comment]
@rpjohnst Just to clarify: You propose that all `&`s of the constituent's type will be undone, and then the implicit `ref` will add one level of `&`, right? 
So matching `(a, b)` against `&(1, &&2)` will desugar to [`&(ref a, &&ref b)`](https://play.rust-lang.org/?gist=013efcdb1e03704adb7439addd6ae38b&version=stable&mode=debug) (which will result in `a: &i32` and `b: &i32`), correct? (Similar with more levels of `&`)
So one could remove the `&` introduced by the implicit ref, too, by writing `(a, &b)`.
The rule that all constituents will only have one level of `&` is easy to remember.

It works well for read-only (`&`) references because for those it makes no difference how many levels of references the type has.
But for mutable references it makes a difference, e.g.:
Matching `(a, b)` against `&mut (1, &mut &mut 2)` will desugar to `&mut (ref mut a, &mut &mut ref mut b)`, resulting in `b: &mut i32` (with that rule).
[But you can do less with a `&mut T` than with a `&mut &mut T`](https://play.rust-lang.org/?gist=30c2a3ab07e7eacd6250abfd77cb8959&version=stable&mode=debug), so in some cases, users might want to keep multiple levels of `&mut`.

So, if we want to preserve the ability to keep all the `&mut`-ref levels of the original type, we have to do it for read-only refs, too (if we want to preserve consistency between read-only refs and mutable refs).

So then we can't auto-remove all the `&` levels of the original type, but we can keep the current rule of `match_default_bindings` that's equally easy to remember ("each constituent's type will get one additional ref") but allow the user to undo each level by using `&` in the pattern. E.g. matching `(_, &&&b)` against `&(1, &&2)` will result in `b: i32` (all `&` levels undone, including the implicit `ref`), and matching `(_, &mut &mut &mut b)` against `&mut (1, &mut &mut 2)` will also result in `b: i32`.

(But I think it makes sense to allow undoing a `&mut` by writing `&`. E.g. matching `(_, &&&b)` against `&mut (1, &mut &mut 2)` to get `b: i32`.)

**Comment by rpjohnst** [comment]
> So matching `(a, b)` against `&(1, &&2)` will desugar to [`&(ref a, &&ref b)`](https://play.rust-lang.org/?gist=013efcdb1e03704adb7439addd6ae38b&version=stable&mode=debug) (which will result in `a: &i32` and b: &i32`), correct? (Similar with more levels of `&`)

Yes, exactly. Kind of doubling down on the idea that `&`s aren't part of the type's structure while in ref mode.

> So, if we want to preserve the ability to keep all the `&mut`-ref levels of the original type, we have to do it for read-only refs, too (if we want to preserve consistency between read-only refs and mutable refs).

IMO we shouldn't do this at all- the move binding mode is sufficient to control the levels of references. If you need that control, just switch back to that mode with a `&` *before* you reach the binding. Straightforwardness and consistency for `&` vs `&mut` for the ref binding modes, and full control for the move binding mode.

The idea of accumulating extra reference levels, as you propose, is basically the thing that we (mis)interpreted @eddyb's version as, and which has the problem of introducing dummy temporaries, and which is backwards-incompatible with the #46688 fix.

**Comment by Boscop** [comment]
@rpjohnst 
> The idea of accumulating extra reference levels, as you propose, is basically the thing that we (mis)interpreted @eddyb's version as, and which has the problem of introducing dummy temporaries, and which is backwards-incompatible with the #46688 fix.

Hm, which dummy temporaries would it introduce?

Wouldn't your proposal also be backwards incompatible (because currently, matching `(_, b)` against `&(1, &2)` results in `b: &&i32` but afterwards, it would result in `b: &i32`)?

**Comment by rpjohnst** [comment]
It introduces dummy temporaries when you have some structure in between the layers of references. If you're just matching against an `&&&i32` or whatever then that's unnecessary.

And yes, it is backwards-incompatible in cases like matching `(_, b)` against `&(1, &2)` or `.. x ..` against `.. &&1 ..`, which is why I mentioned a point release. I'm not sure how else to simultaneously get your proposal working, fix #46688, and not give `&` multiple meanings.

**Comment by Boscop** [comment]
Any update on this issue? :)

**Comment by rpjohnst** [comment]
So I think the place we're at now is @Boscop's original proposal, summarized by @nikomatsakis in https://github.com/rust-lang/rust/issues/42640#issuecomment-386368289 as "make `&P` patterns, when matching a non-reference type *and* in a "default binding mode" of `ref`, go back to a default binding mode of `move`."

While in a sense this gives `&` two meanings, as I expressed in https://github.com/rust-lang/rust/issues/42640#issuecomment-389692171, AFAICT it's the only thing that's backwards-compatible, unlike @eddyb's and my later proposals. So it's probably the best we can do if we want to solve the problem in https://github.com/rust-lang/rust/issues/50008.

I'd like to solve this---it came up again [on Reddit](https://www.reddit.com/r/rust/comments/91iy0n/my_experience_with_the_rust_2018_preview/e2yxp95/) and it's kind of a pain. Is the next step just to write an RFC for it?

**Comment by rpjohnst** [comment]
@withoutboats Ah, I didn't realize this would be in conflict with the second two options, especially in light of https://github.com/rust-lang/rust/pull/48448.

So would the next step be to get consensus on which approach to take? Do we need to stick with just one, or could we have both to round out the existing behavior?

**Comment by withoutboats** [comment]
@rpjohnst yea, we need consensus building to move forward here. as far as i know, we don't even have consensus that we shouldn't do more than one of them.

**Comment by Boscop** [comment]
@withoutboats 

> 3. Finally, it's been argued that another, currently postponed proposal to insert dereferences to convert &T to T where T: Copy would resolve this problem in practice.

Always, unconditionally? But `&mut T` couldn't be unconditionally dereferenced/moved, because the code in the body might want to mutate the original. 

> 2. Another option would be to introduce a new pattern syntax, like (a, move b), instead of (a, &b).

Please don't forget, it should be possible to mark the moved-out copy as `mut` locally, like `&(mut x)` to copy the `&T` as a `T` and make it locally mutable. (Using [the existing `&(mut x)` and `&mut (mut x)` syntax](https://play.rust-lang.org/?gist=0f527172accd24066263428f3565d120&version=nightly&mode=debug&edition=2015).)

> 1. As you said, using the & sigil is one option.

I really think the best option is to use the `&` sigil because that's the most consistent/familiar way and [what people would expect](https://github.com/rust-lang/rust/issues/50008).
And people who know the `&(mut x)` syntax would expect that to work with the `&` sigil, too!
(If we don't use the `&` sigil, we'd also have to special-case another syntax to make the move-copied value locally `mut`!)

Btw, what I would like (in addition to going with the `&` syntax option), is being able to use `&` also for move-copying a `&mut T` (instead of having to write `&mut x`) (this would also allow using `&(mut x)` to move-copy a `&mut T` and make it `mut` in the local scope).

**Comment by rpjohnst** [comment]
Having thought through this some more, this is one place in the language where I can easily see supporting both `&` and `move` like I mentioned above. We *already* have two "modes" and two prominent corresponding mental models- old-style patterns with "`&T` is an address," and ergonomics-style patterns with "`&T` is a less-permissive `T`."

So people (and, more importantly IMO, particular pieces of programs given their context) can use the one that expresses their intent more clearly. When `&` (and later, `Box`/`Rc`/etc) matter as part of the structure of your type, use `&`; when they don't and you just want to talk about permissions, use `ref`/`move`.

Does this make sense to anyone else? Or would people rather stick to just one solution? Or do people like the "coerce `&T` to `T` where `T: Copy`" idea better? In the last case, would such a coercion be *more* than something like autoderef-on-operators (e.g. in argument position), and does that really seem like something we could get consensus on (I would be really surprised)?

**Comment by nikomatsakis** [comment]
This is implemented and stabilizing, but we never closed it! The only real remaining question is https://github.com/rust-lang/rust/issues/44849 and we have a separate issue for that.

**Comment by Boscop** [comment]
@withoutboats @rpjohnst  Any update on this?
I think if we support `move` for this case, we'd also need to allow it in non-match_default_bindings situations for consistency, e.g. `let x = Some(&1); match x { Some(move a) => a, None => 0 }` and `v.iter().map(|move item| ..)` and I wouldn't like that, because then we have 2 competing equivalent ways to express the same thing (which wouldn't be orthogonal)! 
I also don't like the auto-move of `T:Copy` because some `Copy` types are very large, and apart from this, I think the programmer should have a choice, to leave it as a ref or to explicitly move it out. 
I think the `&a` syntax is our best option and I don't think it would be confusing, once people understand that match_default_bindings gives references. The `&a` syntax behaves here just like in `let x = Some(&1); match x { Some(&a) => a, None => 0 }` and `v.iter().map(|&item| ..)`.
I really don't think it'll be a problem, especially because people wouldn't accidentally use the `&a` syntax unless they want to move it out.

**Comment by tamuhey** [comment]
One of the examples in the RFC cannot be built:

```rust
fn main(){
    let x = &Some((3, 3));
    match x {
      // Here, each of the patterns are treated independently
      Some((x, 3)) | &Some((ref x, 5)) => {}
      _ => {}
    }
}
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=73c5447b73e242211f75f9bc9abd62f8)

Why?
